### ISSUE TITLE
Usage of "attrib" keywords in defVar

### ISSUE THREAD

**Issue by natgeo-wong** [Alexander-Barth/NCDatasets.jl#73]
Hi Alexander,

I was using `defVar` to save some data with the format `Int16` into a dataset.  However, I realized that I was using `atts` as a keyword, which is probably a legacy of when I used `NetCDF.jl` to handle NetCDF data in my code.  So I switched to using `attrib` instead.

However, then I ran into the following bug:
```
ERROR: LoadError: InexactError: trunc(Int16, -2.3592272767e10)
Stacktrace:
 [1] trunc at ./float.jl:675 [inlined]
 [2] round at ./float.jl:361 [inlined]
 [3] _broadcast_getindex_evalf at ./broadcast.jl:625 [inlined]
 [4] _broadcast_getindex at ./broadcast.jl:608 [inlined]
 [5] getindex at ./broadcast.jl:558 [inlined]
 [6] macro expansion at ./broadcast.jl:888 [inlined]
 [7] macro expansion at ./simdloop.jl:77 [inlined]
 [8] copyto! at ./broadcast.jl:887 [inlined]
 [9] copyto! at ./broadcast.jl:842 [inlined]
 [10] copy at ./broadcast.jl:818 [inlined]
 [11] materialize at ./broadcast.jl:798 [inlined]
 [12] setindex!(::NCDatasets.Variable{Int16,3}, ::Array{Float64,3}, ::Colon) at /Users/natgeo-wong/.julia/packages/NCDatasets/uW2kc/src/NCDatasets.jl:1069
 [13] setindex!(::NCDatasets.CFVariable{Union{Missing, DateTime, CFTime.AbstractCFDateTime, Number},3,NCDatasets.Variable{Int16,3},NCDatasets.Attributes}, ::Array{Int16,3}, ::Colon) at /Users/natgeo-wong/.julia/packages/NCDatasets/uW2kc/src/NCDatasets.jl:1412
 [14] #defVar#24(::Base.Iterators.Pairs{Symbol,Dict{Any,Any},Tuple{Symbol},NamedTuple{(:attrib,),Tuple{Dict{Any,Any}}}}, ::typeof(NCDatasets.defVar), ::NCDatasets.Dataset, ::SubString{String}, ::Array{Int16,3}, ::Tuple{String,String,String}) at /Users/natgeo-wong/.julia/packages/NCDatasets/uW2kc/src/NCDatasets.jl:738
 [15] (::getfield(NCDatasets, Symbol("#kw##defVar")))(::NamedTuple{(:attrib,),Tuple{Dict{Any,Any}}}, ::typeof(NCDatasets.defVar), ::NCDatasets.Dataset, ::SubString{String}, ::Array{Int16,3}, ::Tuple{String,String,String}) at ./none:0
 [16] clisatrawsave(::Array{Int16,3}, ::Array{Any,1}, ::String, ::Dict{Any,Any}, ::Date) at /Users/natgeo-wong/.julia/dev/ClimateSatellite/src/raw.jl:101
 [17] #gpmdwn#14(::Bool, ::typeof(gpmdwn), ::Array{String,1}, ::Date, ::Dict{Any,Any}) at /Users/natgeo-wong/.julia/dev/ClimateSatellite/src/gpm.jl:166
 [18] #gpmdwn at ./none:0 [inlined]
 [19] #clisatdownload#5(::String, ::String, ::Array{String,1}, ::Bool, ::typeof(clisatdownload), ::String, ::Date) at /Users/natgeo-wong/.julia/dev/ClimateSatellite/src/raw.jl:30
 [20] (::getfield(ClimateSatellite, Symbol("#kw##clisatdownload")))(::NamedTuple{(:email, :regions),Tuple{String,Array{String,1}}}, ::typeof(clisatdownload), ::String, ::Date) at ./none:0
 [21] top-level scope at /Users/natgeo-wong/Projects/JuliaClimate/clisat_download/gpm_test.jl:6
in expression starting at /Users/natgeo-wong/Projects/JuliaClimate/clisat_download/gpm_test.jl:6
```

It appears that using the `attrib` keyword actually might be applying the `scale` and `add_offset` keyword arguments when saving the data, which is not what I wanted (see the truncation error above).

Am I doing the attributes correctly?
```
defVar(ds,var_var[ii],data,("longitude","latitude","time"),attrib=att_var[ii]);
```

where `var_var[ii]` is the i-th variable name, `data` is the Int16 data, and `att_var[ii]` contains the attributes for the i-th variable?  `att_var` is defined as

```
    for ii = 1 : nvar
        var_var[ii]                  = info["varID"][ii];
        att_var[ii]["units"]         = info["units"][ii];
        att_var[ii]["standard_name"] = info["standard"][ii];
        att_var[ii]["long_name"]     = info["variable"][ii];
        att_var[ii]["scale_factor"]  = info["scale"][ii];
        att_var[ii]["add_offset"]    = info["offset"][ii];
        att_var[ii]["missing_value"] = -32768;
    end
```

**Comment by natgeo-wong** [comment]
I tried doing

```
defVar(ds,var_var[1],data,("longitude","latitude","time"),
               attrib=["units" => info["units"][1],
                       "standard_name" => info["standard"][1],
                       "long_name" => info["variable"][1],
                       "scale_factor" => info["scale"][1],
                       "add_offset" => info["offset"][1],
                       "missing_value" => -32768]);
```

As was shown using an example in the documentation, but this just gives me

```
ERROR: LoadError: UndefRefError: access to undefined reference
Stacktrace:
 [1] getindex(::Array{AbstractString,1}, ::Int64) at ./array.jl:728
 [2] clisatrawsave(::Array{Int16,3}, ::Array{Any,1}, ::String, ::Dict{Any,Any}, ::Date) at /Users/natgeo-wong/.julia/dev/ClimateSatellite/src/raw.jl:95
 [3] #gpmdwn#12(::Bool, ::typeof(gpmdwn), ::Array{String,1}, ::Date, ::Dict{Any,Any}) at /Users/natgeo-wong/.julia/dev/ClimateSatellite/src/gpm.jl:166
 [4] #gpmdwn at ./none:0 [inlined]
 [5] #clisatdownload#5(::String, ::String, ::Array{String,1}, ::Bool, ::typeof(clisatdownload), ::String, ::Date) at /Users/natgeo-wong/.julia/dev/ClimateSatellite/src/raw.jl:30
 [6] (::getfield(ClimateSatellite, Symbol("#kw##clisatdownload")))(::NamedTuple{(:email, :regions),Tuple{String,Array{String,1}}}, ::typeof(clisatdownload), ::String, ::Date) at ./none:0
 [7] top-level scope at /Users/natgeo-wong/Projects/JuliaClimate/clisat_download/gpm_test.jl:6
in expression starting at /Users/natgeo-wong/Projects/JuliaClimate/clisat_download/gpm_test.jl:6
```

**Comment by Alexander-Barth** [comment]
Does it work when you use OrderedDict or Dict for a list of attributes? It could be this issue described here:
https://alexander-barth.github.io/NCDatasets.jl/latest/attributes/#Possible-type-promotion-in-Julia-1 ?

Do you see the issue template when you file an issue? 
https://github.com/Alexander-Barth/NCDatasets.jl/blob/master/.github/ISSUE_TEMPLATE.md

In particular a minimal working reproducer is important for me to debug such issues.

**Comment by natgeo-wong** [comment]
Okay, sorry about that, let me rephrase into the template given.

**Describe the bug**

I don't know that this is so much a bug, as it is a difference in personal workflow.

Basically, when I downloaded satellite data, the data is in Float64 format.  For compression purposes however, I want to save in Int16 format.  I have therefore converted the data from Float64 into Int16 format independently of NCDatasets.jl.

However, when I try to save the data into a NetCDF file, with the attribs `scale_factor` and `add_offset`, the attribs are immediately applied to the data a second time, which is not what I wanted because I have already converted the data into the required format.

**To Reproduce**

```
testdata = Int16.(ones(23,34,13)*32767); att_var = Dict();
att_var["scale_factor"] = 1.38888888888889e-6   
att_var["add_offset"] = 0.0455097222222222
ds = Dataset("test.nc","c");
ds.dim["longitude"],ds.dim["latitude"],ds.dim["time"] = size(testdata);
defVar(ds,"testdata",testdata,("longitude","latitude","time"),attrib=att_var);
```

And this gives the error

```
ERROR: InexactError: trunc(Int16, 2.3592207233e10)
Stacktrace:
 [1] trunc at ./float.jl:675 [inlined]
 [2] round at ./float.jl:361 [inlined]
 [3] _broadcast_getindex_evalf at ./broadcast.jl:625 [inlined]
 [4] _broadcast_getindex at ./broadcast.jl:608 [inlined]
 [5] getindex at ./broadcast.jl:558 [inlined]
 [6] macro expansion at ./broadcast.jl:888 [inlined]
 [7] macro expansion at ./simdloop.jl:77 [inlined]
 [8] copyto! at ./broadcast.jl:887 [inlined]
 [9] copyto! at ./broadcast.jl:842 [inlined]
 [10] copy at ./broadcast.jl:818 [inlined]
 [11] materialize at ./broadcast.jl:798 [inlined]
 [12] setindex!(::NCDatasets.Variable{Int16,3}, ::Array{Float64,3}, ::Colon) at /Users/natgeo-wong/.julia/packages/NCDatasets/uW2kc/src/NCDatasets.jl:1069
 [13] setindex!(::NCDatasets.CFVariable{Union{Missing, DateTime, AbstractCFDateTime, Number},3,NCDatasets.Variable{Int16,3},NCDatasets.Attributes}, ::Array{Int16,3}, ::Colon) at /Users/natgeo-wong/.julia/packages/NCDatasets/uW2kc/src/NCDatasets.jl:1412
 [14] #defVar#24(::Base.Iterators.Pairs{Symbol,Dict{Any,Any},Tuple{Symbol},NamedTuple{(:attrib,),Tuple{Dict{Any,Any}}}}, ::typeof(defVar), ::Dataset, ::String, ::Array{Int16,3}, ::Tuple{String,String,String}) at /Users/natgeo-wong/.julia/packages/NCDatasets/uW2kc/src/NCDatasets.jl:738
 [15] (::getfield(NCDatasets, Symbol("#kw##defVar")))(::NamedTuple{(:attrib,),Tuple{Dict{Any,Any}}}, ::typeof(defVar), ::Dataset, ::String, ::Array{Int16,3}, ::Tuple{String,String,String}) at ./none:0
 [16] top-level scope at none:0
```

**Expected behavior**

What I actually want to do is save the variable `testdata` as is into the NetCDF files, without applying the `scale_factor` and `add_offset` attributes in the process.

My current workaround for this (which I just discovered in passing yesterday after having a brainwave) is to do `v = defVar()` and then do `v.var[:] = testdata`, but to me this is slightly counterintuitive.

I understand the rationale behind applying `scale_factor` and `add_offset` when saving the dataset in order to save time, but the problem for me was that the `scale_factor` and `add_offset` I applied to convert my raw data into Int16 is different from the `scale_factor` and `add_offset` I need to apply to the dataset in order to satisfy CF conventions.

**Comment by Alexander-Barth** [comment]
> Okay, sorry about that, let me rephrase into the template given.

No problem, thanks a lot for revising your issue and providing this information :-)

Yes, your values are too large to hold into a Int16 when you applied the convertions:

```
 (maximum(testdata) - 0.0455097222222222)/ 1.38888888888889e-6
2.359220723299998e10

julia> typemax(Int16)
32767
```
> My current workaround for this (which I just discovered in passing yesterday after having a brainwave) is to do v = defVar() and then do v.var[:] = testdata, but to me this is slightly counterintuitive.

This is actually the indented bahviour:
* Use `v[:]` to apply `scale_factor`/`add_offset` (reading or writing)
* Use `v.var[:]` if you do not want to apply `scale_factor`/`add_offset` (reading or writing)

> but the problem for me was that the scale_factor and add_offset I applied to convert my raw data into Int16 is different from the scale_factor and add_offset I need to apply to the dataset in order to satisfy CF conventions.

This is not clear to me. It is the inverse operation:
* when reading:  data = scale_factor * raw + add_offset
* when writing:  raw =  (data - add_offset) / scale_factor

**Comment by natgeo-wong** [comment]
I think I just come from a workframe where I tend to save data into NetCDF as-is, without applying any `scale_factor / add_offset` to the data, so sometimes the tricks of NetCDF tend to trip me up slightly.

Thanks so much for helping!

**Comment by Alexander-Barth** [comment]
You are welcome!
